File: packages/hushspec/src/adapters/mcp.ts
import type { EvaluationAction, EvaluationResult } from '../evaluate.js';
import type { HushGuard } from '../middleware.js';

export function mapMCPToolCall(
  toolName: string,
  args?: Record<string, unknown>,
): EvaluationAction {
  const mappings: Record<
    string,
    (a?: Record<string, unknown>) => EvaluationAction
  > = {
    read_file: (a) => ({
      type: 'file_read',
      target: (a?.path as string) ?? '',
    }),
    write_file: (a) => ({
      type: 'file_write',
      target: (a?.path as string) ?? '',
      content: a?.content as string,
    }),
    list_directory: (a) => ({
      type: 'file_read',
      target: (a?.path as string) ?? '',
    }),
    run_command: (a) => ({
      type: 'shell_command',
      target: (a?.command as string) ?? '',
    }),
    execute: (a) => ({
      type: 'shell_command',
      target: (a?.command as string) ?? '',
    }),
    fetch: (a) => ({
      type: 'egress',
      target: extractDomain((a?.url as string) ?? ''),
    }),
    http_request: (a) => ({
      type: 'egress',
      target: extractDomain((a?.url as string) ?? ''),
    }),
  };

  const mapper = mappings[toolName];
  if (mapper) return mapper(args);

  return {
    type: 'tool_call',
    target: toolName,
    args_size: args ? JSON.stringify(args).length : undefined,
  };
}

export function extractDomain(url: string): string {
  try {
    return new URL(url).hostname;
  } catch {
    return url;
  }
}

export function createMCPGuard(
  guard: HushGuard,
): (toolName: string, args?: Record<string, unknown>) => EvaluationResult {
  return (
    toolName: string,
    args?: Record<string, unknown>,
  ): EvaluationResult => {
    const action = mapMCPToolCall(toolName, args);
    return guard.evaluate(action);
  };
}

File: packages/python/hushspec/adapters/mcp.py
from __future__ import annotations

import json
from typing import Optional
from urllib.parse import urlparse

from hushspec.evaluate import EvaluationAction, EvaluationResult
from hushspec.middleware import HushGuard


def extract_domain(url: str) -> str:
    try:
        parsed = urlparse(url)
        return parsed.hostname or url
    except Exception:
        return url


def map_mcp_tool_call(
    tool_name: str,
    args: Optional[dict] = None,
) -> EvaluationAction:
    args = args or {}

    mappings = {
        "read_file": lambda a: EvaluationAction(
            type="file_read",
            target=a.get("path", ""),
        ),
        "write_file": lambda a: EvaluationAction(
            type="file_write",
            target=a.get("path", ""),
            content=a.get("content"),
        ),
        "list_directory": lambda a: EvaluationAction(
            type="file_read",
            target=a.get("path", ""),
        ),
        "run_command": lambda a: EvaluationAction(
            type="shell_command",
            target=a.get("command", ""),
        ),
        "execute": lambda a: EvaluationAction(
            type="shell_command",
            target=a.get("command", ""),
        ),
        "fetch": lambda a: EvaluationAction(
            type="egress",
            target=extract_domain(a.get("url", "")),
        ),
        "http_request": lambda a: EvaluationAction(
            type="egress",
            target=extract_domain(a.get("url", "")),
        ),
    }

    mapper = mappings.get(tool_name)
    if mapper is not None:
        return mapper(args)

    return EvaluationAction(
        type="tool_call",
        target=tool_name,
        args_size=len(json.dumps(args)) if args else None,
    )


def create_mcp_guard(guard: HushGuard):

    def handler(
        tool_name: str,
        args: Optional[dict] = None,
    ) -> EvaluationResult:
        action = map_mcp_tool_call(tool_name, args)
        return guard.evaluate(action)

    return handler
